internal/kms/azure_vault.go

```diff
@@ -0,0 +1,230 @@
+}
+
+func (kms *azureKeyVaultKMS) Destroy() {
+	// Nothing to do.
+}
+
+func (kms *azureKeyVaultKMS) getService() (*azsecrets.Client, error) {
+	certs, key, err := azidentity.ParseCertificates([]byte(kms.clientCertificate), []byte{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse Azure client certificate: %w", err)
+	}
+	creds, err := azidentity.NewClientCertificateCredential(kms.tenantID, kms.clientID, certs, key, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Azure credentials: %w", err)
+	}
+
+	azClient, err := azsecrets.NewClient(kms.vaultURL, creds, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Azure client: %w", err)
+	}
+
+	return azClient, nil
+}
+
+func (kms *azureKeyVaultKMS) getSecrets() (map[string]interface{}, error) {
+	c, err := k8s.NewK8sClient()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to kubernetes to "+
+			"get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	secret, err := c.CoreV1().Secrets(kms.namespace).Get(context.TODO(),
+		kms.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w", kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
+	for k, v := range secret.Data {
+		switch k {
+		case azureClientCertificate:
+			config[k] = string(v)
+		default:
+			return nil, fmt.Errorf("unsupported option for KMS provider %q: %s", kmsTypeAzure, k)
+		}
+	}
+
+	return config, nil
+}
+
+// FetchDEK returns passphrase from Vault.
+func (kms *azureKeyVaultKMS) FetchDEK(key string) (string, error) {
+	svc, err := kms.getService()
+	if err != nil {
+		return "", fmt.Errorf("failed to get KMS service: %w", err)
+	}
+	// Since the second return variable loss.Version is not used, there it is ignored.
+	getSecretResponse, err := svc.GetSecret(context.TODO(), key, "", nil)
+	if err != nil {
+		return "", fmt.Errorf("failed to get secret: %w", err)
+	}
+
+	return *getSecretResponse.Value, nil
+}
+
+// StoreDEK saves new passphrase in Vault.
+func (kms *azureKeyVaultKMS) StoreDEK(key, value string) error {
+	svc, err := kms.getService()
+	if err != nil {
+		return fmt.Errorf("failed to get KMS service: %w", err)
+	}
+
+	setSecretParams := azsecrets.SetSecretParameters{
+		Value: &value,
+	}
+	_, err = svc.SetSecret(context.TODO(), key, setSecretParams, nil)
+	if err != nil {
+		return fmt.Errorf("failed to set seceret %w", err)
+	}
+
+	return nil
+}
+
+// RemoveDEK deletes passphrase from Vault.
+func (kms *azureKeyVaultKMS) RemoveDEK(key string) error {
+	svc, err := kms.getService()
+	if err != nil {
+		return fmt.Errorf("failed to get KMS service: %w", err)
+	}
+
+	_, err = svc.DeleteSecret(context.TODO(), key, nil)
+	if err != nil {
+		return fmt.Errorf("failed to delete seceret %w", err)
+	}
+
+	// The purge deleted secret operation removes the secret permanently.
+	attempts := 3
+	attemptDelay := 1 * time.Second
+
+	for i := 0; i < attempts; i++ {
```

Comment:
why do we need to try 3 times with sleep? why cant we just return error, if there is any failure and let it retried as part of RPC call not like this?

Reply:
`DeleteSecret()` takes few seconds to delete the secret  and calling `PurgeDeletedSecret()` will fail if secret is still being deleted. I thought retry with few delay should be okay. 
But yea, as you said this may not be a good way.


Currently, we just log warning if `RemoveDEK()` fails and continue with deleting image.
https://github.com/ceph/ceph-csi/blob/328e4e5a0fa06521c5508b94a5b7ff250ada1681/internal/rbd/rbd_util.go#L636-L641

other way what you suggest is return an error if `RemoveDEK()` fails and let RPC make a retry. 
In this case, image won't be deleted until the secret is deleted from the kms?

Reply:
i would go with the option to let RPC retry and let the image be present until the key is deleted, you can return in progress error in RPC to show that deletion is in progress.

Reply:
or if we are fine with leaving the keys garbage in KMS is also fine, just delete it and go for deleting image. 
Lets hear from others what they think is the right approach.

cc @nixpanic @Rakshith-R

Reply:
yes, Azure has retention policy to be configured between 7 to 90 days.
probably, now I think we don't need do purging.
Thanks @nixpanic @Madhu-1